domain/network/state/subnet.go

```diff
@@ -0,0 +1,397 @@
+    fan_subnet.cidr                      AS &Subnet.underlay_cidr,
+    availability_zone.name               AS &Subnet.az,
+    provider_space.provider_id           AS &Subnet.provider_space_uuid
+FROM subnet 
+    LEFT JOIN fan_subnet 
+    ON subnet.uuid = fan_subnet.subject_subnet_uuid
+    JOIN provider_subnet
+    ON subnet.uuid = provider_subnet.subnet_uuid
+    JOIN provider_network_subnet
+    ON subnet.uuid = provider_network_subnet.subnet_uuid
+    JOIN provider_network
+    ON provider_network_subnet.provider_network_uuid = provider_network.uuid
+    LEFT JOIN availability_zone_subnet
+    ON availability_zone_subnet.subnet_uuid = subnet.uuid
+    LEFT JOIN availability_zone
+    ON availability_zone_subnet.availability_zone_uuid = availability_zone.uuid
+    LEFT JOIN provider_space
+    ON subnet.space_uuid = provider_space.space_uuid`
+
+// GetAllSubnets returns all known subnets in the model.
+func (st *State) GetAllSubnets(
+	ctx context.Context,
+) (network.SubnetInfos, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Append the space uuid condition to the query only if it's passed to the function.
+	q := retrieveSubnetsStmt + ";"
+
+	s, err := sqlair.Prepare(q, Subnet{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var rows Subnets
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying subnets")
+	}
+
+	return rows.ToSubnetInfos(), nil
+}
+
+// GetSubnet returns the subnet by UUID.
+func (st *State) GetSubnet(
+	ctx context.Context,
+	uuid utils.UUID,
+) (*network.SubnetInfo, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Append the space uuid condition to the query only if it's passed to the function.
+	q := retrieveSubnetsStmt + " WHERE subnet.uuid = $M.id;"
+
+	s, err := sqlair.Prepare(q, Subnet{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var rows Subnets
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{"id": uuid.String()}).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying subnets")
+	}
+
+	if len(rows) == 0 {
+		return nil, errors.NotFoundf("subnet %q", uuid.String())
+	}
+
+	return &rows.ToSubnetInfos()[0], nil
+}
+
+func updateSubnetSpaceIDTx(
+	ctx context.Context,
+	tx *sql.Tx,
+	uuid string,
+	spaceID string,
+) error {
+	q := `
+UPDATE subnet
+SET    space_uuid = ?
+WHERE  uuid = ?;`
+
+	rows, err := tx.ExecContext(ctx, q, spaceID, uuid)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	affected, err := rows.RowsAffected()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if affected != 1 {
+		return errors.NotFoundf("subnet %q", uuid)
+	}
+
+	return nil
+}
```

Comment:
Does this warrant a separate function if it's only being used in two places? Just a thought...

Reply:
I prefer to have the function defined here (in subnet.go) since it's an update on the subnet table, and then it's used on the space.go file. Indeed I could have the query repeated in both places, I prefer it this way but I can change it if you have strong opinions on this.